packages/tailwindcss/src/variants.ts

```diff
@@ -179,6 +263,35 @@ export class Variants {
   }
 }
 
+export function compoundsForSelectors(selectors: string[]) {
+  let compounds = Compounds.Never
+
+  for (let sel of selectors) {
+    if (sel[0] === '@') {
+      // Non-conditional at-rules are present so we can't compound
+      if (
+        !sel.startsWith('@media') &&
+        !sel.startsWith('@supports') &&
+        !sel.startsWith('@container')
+      ) {
+        return Compounds.Never
+      }
+
+      compounds |= Compounds.AtRules
+      continue
+    }
+
+    // Pseudo-elements are present so we can't compound
+    if (sel.includes('::')) {
```

Comment:
Do we have to think about `::file-selector-button` as special here? I have tested a few things and don't _think_ so, because in my testing `:not(::file-selector-button)` doesn't work in any browser even though stuff like `::file-selector-button:is(:not(:active))` does work.

Reply:
No we don't — pseudo-elements don't work in :not(…), :is(…), :where(…), :has(…) and can't appear in the middle of a selector so `.group` and `.peer` are out too.

As for your 2nd example I think that's `file:not-active:{utility}` yeah?